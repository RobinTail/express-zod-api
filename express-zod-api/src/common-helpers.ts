import { Request } from "express";
import * as R from "ramda";
import { z } from "zod";
import { CommonConfig, InputSource, InputSources } from "./config-type";
import { contentTypes } from "./content-type";
import { OutputValidationError } from "./errors";
import { metaSymbol } from "./metadata";
import { AuxMethod, Method } from "./method";

/** @desc this type does not allow props assignment, but it works for reading them when merged with another interface */
export type EmptyObject = Record<string, never>;
export type EmptySchema = z.ZodObject<EmptyObject, "strip">;
export type FlatObject = Record<string, unknown>;

/** @link https://stackoverflow.com/a/65492934 */
type NoNever<T, F> = [T] extends [never] ? F : T;

/**
 * @desc Using module augmentation approach you can specify tags as the keys of this interface
 * @example declare module "express-zod-api" { interface TagOverrides { users: unknown } }
 * @link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 * */
// eslint-disable-next-line @typescript-eslint/no-empty-object-type -- augmentation
export interface TagOverrides {}
export type Tag = NoNever<keyof TagOverrides, string>;

/** @see https://expressjs.com/en/guide/routing.html */
export const routePathParamsRegex = /:([A-Za-z0-9_]+)/g;
export const getRoutePathParams = (path: string): string[] =>
  path.match(routePathParamsRegex)?.map((param) => param.slice(1)) || [];

const areFilesAvailable = (request: Request): boolean => {
  const contentType = request.header("content-type") || "";
  const isUpload = contentType.toLowerCase().startsWith(contentTypes.upload);
  return "files" in request && isUpload;
};

export const defaultInputSources: InputSources = {
  get: ["query", "params"],
  post: ["body", "params", "files"],
  put: ["body", "params"],
  patch: ["body", "params"],
  delete: ["query", "params"],
};
const fallbackInputSource: InputSource[] = ["body", "query", "params"];

export const getActualMethod = (request: Request) =>
  request.method.toLowerCase() as Method | AuxMethod;

export const getInput = (
  req: Request,
  userDefined: CommonConfig["inputSources"] = {},
): FlatObject => {
  const method = getActualMethod(req);
  if (method === "options") return {};
  return (
    userDefined[method] ||
    defaultInputSources[method] ||
    fallbackInputSource
  )
    .filter((src) => (src === "files" ? areFilesAvailable(req) : true))
    .reduce<FlatObject>((agg, src) => Object.assign(agg, req[src]), {});
};

export const ensureError = (subject: unknown): Error =>
  subject instanceof Error ? subject : new Error(String(subject));

export const getMessageFromError = (error: Error): string => {
  if (error instanceof z.ZodError) {
    return error.issues
      .map(({ path, message }) =>
        (path.length ? [path.join("/")] : []).concat(message).join(": "),
      )
      .join("; ");
  }
  if (error instanceof OutputValidationError) {
    const hasFirstField = error.cause.issues[0]?.path.length > 0;
    return `output${hasFirstField ? "/" : ": "}${error.message}`;
  }
  return error.message;
};

/** Takes the original unvalidated examples from the properties of ZodObject schema shape */
export const pullExampleProps = <T extends z.SomeZodObject>(subject: T) =>
  Object.entries(subject.shape).reduce<Partial<z.input<T>>[]>(
    (acc, [key, schema]) => {
      const { _def } = schema as z.ZodType;
      return combinations(
        acc,
        (_def[metaSymbol]?.examples || []).map(R.objOf(key)),
        ([left, right]) => ({ ...left, ...right }),
      );
    },
    [],
  );

export const getExamples = <
  T extends z.ZodType,
  V extends "original" | "parsed" | undefined,
>({
  schema,
  variant = "original",
  validate = variant === "parsed",
  pullProps = false,
}: {
  schema: T;
  /**
   * @desc examples variant: original or parsed
   * @example "parsed" — for the case when possible schema transformations should be applied
   * @default "original"
   * @override validate: variant "parsed" activates validation as well
   * */
  variant?: V;
  /**
   * @desc filters out the examples that do not match the schema
   * @default variant === "parsed"
   * */
  validate?: boolean;
  /**
   * @desc should pull examples from properties — applicable to ZodObject only
   * @default false
   * */
  pullProps?: boolean;
}): ReadonlyArray<V extends "parsed" ? z.output<T> : z.input<T>> => {
  let examples = schema._def[metaSymbol]?.examples || [];
  if (!examples.length && pullProps && schema instanceof z.ZodObject)
    examples = pullExampleProps(schema);
  if (!validate && variant === "original") return examples;
  const result: Array<z.input<T> | z.output<T>> = [];
  for (const example of examples) {
    const parsedExample = schema.safeParse(example);
    if (parsedExample.success)
      result.push(variant === "parsed" ? parsedExample.data : example);
  }
  return result;
};

export const combinations = <T>(
  a: T[],
  b: T[],
  merge: (pair: [T, T]) => T,
): T[] => (a.length && b.length ? R.xprod(a, b).map(merge) : a.concat(b));

/**
 * @desc isNullable() and isOptional() validate the schema's input
 * @desc They always return true in case of coercion, which should be taken into account when depicting response
 */
export const hasCoercion = (schema: z.ZodTypeAny): boolean =>
  "coerce" in schema._def && typeof schema._def.coerce === "boolean"
    ? schema._def.coerce
    : false;

export const ucFirst = (subject: string) =>
  subject.charAt(0).toUpperCase() + subject.slice(1).toLowerCase();

export const makeCleanId = (...args: string[]) => {
  const byAlpha = R.chain((entry) => entry.split(/[^A-Z0-9]/gi), args);
  const byWord = R.chain(
    (entry) =>
      entry.replaceAll(/[A-Z]+/g, (beginning) => `/${beginning}`).split("/"),
    byAlpha,
  );
  return byWord.map(ucFirst).join("");
};

export const getTransformedType = R.tryCatch(
  <T>(schema: z.ZodEffects<z.ZodTypeAny, unknown, T>, sample: T) =>
    typeof schema.parse(sample),
  R.always(undefined),
);

/** @desc can still be an array, use Array.isArray() or rather R.type() to exclude that case */
export const isObject = (subject: unknown) =>
  typeof subject === "object" && subject !== null;

export const isProduction = R.memoizeWith(
  () => process.env.TSUP_STATIC as string, // eslint-disable-line no-restricted-syntax -- substituted by TSUP
  () => process.env.NODE_ENV === "production", // eslint-disable-line no-restricted-syntax -- memoized
);
