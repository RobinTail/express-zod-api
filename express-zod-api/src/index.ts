import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
