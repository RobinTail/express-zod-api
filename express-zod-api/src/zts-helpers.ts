import type ts from "typescript";
import type { FlatObject } from "./common-helpers.ts";
import type { SchemaHandler } from "./schema-walker.ts";

export interface ZTSContext extends FlatObject {
  isResponse: boolean;
  makeAlias: (key: object, produce: () => ts.TypeNode) => ts.TypeNode;
}

export type Producer = SchemaHandler<ts.TypeNode, ZTSContext>;
