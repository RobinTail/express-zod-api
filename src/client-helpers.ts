import ts from "typescript";

export const f = ts.factory;

export const exportModifier = [f.createModifier(ts.SyntaxKind.ExportKeyword)];

export const publicReadonlyModifier = [
  f.createModifier(ts.SyntaxKind.PublicKeyword),
  f.createModifier(ts.SyntaxKind.ReadonlyKeyword),
];

export const protectedReadonlyModifier = [
  f.createModifier(ts.SyntaxKind.ProtectedKeyword),
  f.createModifier(ts.SyntaxKind.ReadonlyKeyword),
];

const emptyPrefix = f.createTemplateHead("");

const emptyEnding = f.createTemplateTail("");

const spacingSuffix = f.createTemplateMiddle(" ");

export const makeTemplate = (names: (ts.Identifier | string)[]) =>
  f.createTemplateLiteralType(
    emptyPrefix,
    names.map((name, index) =>
      f.createTemplateLiteralTypeSpan(
        f.createTypeReferenceNode(name),
        index === names.length - 1 ? emptyEnding : spacingSuffix
      )
    )
  );

export const parametricIndexNode = makeTemplate(["M", "P"]);

export const makeParam = (
  name: string,
  type?: ts.TypeNode,
  mod?: ts.Modifier[]
) =>
  f.createParameterDeclaration(
    undefined,
    mod,
    undefined,
    name,
    undefined,
    type
  );

export const makeParams = (
  params: Record<string, ts.TypeNode | undefined>,
  mod?: ts.Modifier[]
) =>
  Object.keys(params).reduce(
    (acc, name) => acc.concat(makeParam(name, params[name], mod)),
    [] as ts.ParameterDeclaration[]
  );

export const makeRecord = (
  key: ts.Identifier | ts.KeywordTypeSyntaxKind,
  value: ts.KeywordTypeSyntaxKind
) =>
  f.createExpressionWithTypeArguments(f.createIdentifier("Record"), [
    typeof key === "number"
      ? f.createKeywordTypeNode(key)
      : f.createTypeReferenceNode(key),
    f.createKeywordTypeNode(value),
  ]);

export const makeEmptyInitializingConstructor = (
  params: ts.ParameterDeclaration[]
) => f.createConstructorDeclaration(undefined, params, f.createBlock([]));

export const makeQuotedProp = (name: string, ref: string) =>
  f.createPropertySignature(
    undefined,
    `"${name}"`,
    undefined,
    f.createTypeReferenceNode(ref)
  );

export const makeConst = (name: string, value: ts.Expression) =>
  f.createVariableDeclarationList(
    [f.createVariableDeclaration(name, undefined, undefined, value)],
    ts.NodeFlags.Const
  );

export const makePublicLiteralType = (name: string, literals: string[]) =>
  f.createTypeAliasDeclaration(
    undefined,
    exportModifier,
    name,
    undefined,
    f.createUnionTypeNode(
      literals.map((option) =>
        f.createLiteralTypeNode(f.createStringLiteral(option))
      )
    )
  );

export const makePublicType = (name: string, value: ts.TypeNode) =>
  f.createTypeAliasDeclaration(
    undefined,
    exportModifier,
    name,
    undefined,
    value
  );

export const makePublicReadonlyProp = (
  name: string,
  type: ts.TypeNode,
  exp: ts.Expression
) =>
  f.createPropertyDeclaration(
    publicReadonlyModifier,
    name,
    undefined,
    type,
    exp
  );

export const makePublicClass = (
  name: string,
  constructor: ts.ConstructorDeclaration,
  props: ts.PropertyDeclaration[] = []
) =>
  f.createClassDeclaration(exportModifier, name, undefined, undefined, [
    constructor,
    ...props,
  ]);

export const makeIndexedPromise = (type: ts.Identifier, index: ts.TypeNode) =>
  f.createTypeReferenceNode("Promise", [
    f.createIndexedAccessTypeNode(f.createTypeReferenceNode(type), index),
  ]);

export const makeAnyPromise = () =>
  f.createTypeReferenceNode("Promise", [
    f.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
  ]);

export const makePublicExtendedInterface = (
  name: string,
  extender: ts.HeritageClause[],
  props: ts.PropertySignature[]
) =>
  f.createInterfaceDeclaration(
    exportModifier,
    name,
    undefined,
    extender,
    props
  );

export const makeTypeParams = (params: Record<string, ts.Identifier>) =>
  Object.keys(params).reduce(
    (acc, name) =>
      acc.concat(
        f.createTypeParameterDeclaration(
          [],
          name,
          f.createTypeReferenceNode(params[name])
        )
      ),
    [] as ts.TypeParameterDeclaration[]
  );

export const makeImplementationCallFn = (
  params: string[],
  args: ts.Expression[]
) =>
  f.createArrowFunction(
    undefined,
    undefined,
    params.map((key) => makeParam(key)),
    undefined,
    undefined,
    f.createCallExpression(
      f.createPropertyAccessExpression(f.createThis(), "implementation"),
      undefined,
      args
    )
  );

export const makeObjectKeysReducer = (
  obj: string,
  exp: ts.Expression,
  initial: ts.Expression
) =>
  f.createCallExpression(
    f.createPropertyAccessExpression(
      f.createCallExpression(
        f.createPropertyAccessExpression(f.createIdentifier("Object"), "keys"),
        undefined,
        [f.createIdentifier(obj)]
      ),
      "reduce"
    ),
    undefined,
    [
      f.createArrowFunction(
        undefined,
        undefined,
        makeParams({ acc: undefined, key: undefined }),
        undefined,
        undefined,
        exp
      ),
      initial,
    ]
  );

export const cleanId = (path: string, method: string, suffix: string) => {
  return [method]
    .concat(path.split("/"))
    .concat(suffix)
    .map((entry) => entry.replace(/[^A-Z0-9]/i, ""))
    .map(
      (entry) => entry.slice(0, 1).toUpperCase() + entry.slice(1).toLowerCase()
    )
    .join("");
};
