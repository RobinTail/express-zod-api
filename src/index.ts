import "./zod-plugin";

export { createConfig } from "./config-type";
export { AbstractEndpoint } from "./endpoint";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory";
export {
  getExamples,
  getMessageFromError,
  getStatusCodeFromError,
} from "./common-helpers";
export { createLogger } from "./logger";
export { createMiddleware } from "./middleware";
export {
  createResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler";
export { DependsOnMethod } from "./depends-on-method";
export { ServeStatic } from "./serve-static";
export { createServer, attachRouting } from "./server";
export { Documentation } from "./documentation";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors";
export { testEndpoint } from "./testing";
export { Integration } from "./integration";

export { ez } from "./proprietary-schemas";

// Convenience types
export type { Depicter } from "./documentation-helpers";
export type { Producer } from "./zts-helpers";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
export type { MockOverrides } from "./testing";
export type { Routing } from "./routing";
export type { LoggerOverrides } from "./logger";
export type { FlatObject } from "./common-helpers";
export type { Method } from "./method";
export type { IOSchema } from "./io-schema";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type";
export type { MiddlewareDefinition } from "./middleware";
export type { ResultHandlerDefinition } from "./result-handler";
export type { ApiResponse } from "./api-response";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  CustomHeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security";
