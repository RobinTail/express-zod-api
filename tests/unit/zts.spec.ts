/**
 * This file is based on https://github.com/sachinraja/zod-to-ts
 *
 * MIT License
 *
 * Copyright (c) 2021 Sachin Raja
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import ts from "typescript";
import { z } from "zod";
import { f } from "../../src/client-helpers";
import { defaultSerializer } from "../../src/common-helpers";
import { zodToTs } from "../../src/zts";
import { createTypeAlias, printNode } from "../../src/zts-helpers";

describe("zod-to-ts", () => {
  const printNodeTest = (node: ts.Node) =>
    printNode(node, { newLine: ts.NewLineKind.LineFeed });
  const defaultCtx = {
    isResponse: false,
    getAlias: jest.fn((name: string) => f.createTypeReferenceNode(name)),
    makeAlias: jest.fn(),
    serializer: defaultSerializer,
  };

  describe("z.array()", () => {
    it("outputs correct typescript", () => {
      const node = zodToTs({
        schema: z.object({ id: z.number(), value: z.string() }).array(),
        ...defaultCtx,
      });
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("createTypeAlias()", () => {
    const identifier = "User";
    const node = zodToTs({
      schema: z.object({ username: z.string(), age: z.number() }),
      ...defaultCtx,
    });

    it("outputs correct typescript", () => {
      const typeAlias = createTypeAlias(node, identifier);
      expect(printNodeTest(typeAlias)).toMatchSnapshot();
    });

    it("optionally takes a comment", () => {
      const typeAlias = createTypeAlias(node, identifier, "A basic user");
      expect(printNodeTest(typeAlias)).toMatchSnapshot();
    });
  });

  describe("enums", () => {
    // noinspection JSUnusedGlobalSymbols
    enum Color {
      Red,
      Green,
      Blue,
    }

    // noinspection JSUnusedGlobalSymbols
    enum Fruit {
      Apple = "apple",
      Banana = "banana",
      Cantaloupe = "cantaloupe",
    }

    // noinspection JSUnusedGlobalSymbols
    enum StringLiteral {
      "Two Words",
      "'Quotes\"",
      '\\"Escaped\\"',
    }

    it.each([
      { schema: z.nativeEnum(Color), feature: "numeric" },
      { schema: z.nativeEnum(Fruit), feature: "string" },
      { schema: z.nativeEnum(StringLiteral), feature: "quoted string" },
    ])("handles $feature literals", ({ schema }) => {
      expect(
        printNodeTest(zodToTs({ schema, ...defaultCtx }))
      ).toMatchSnapshot();
    });
  });

  describe("Example", () => {
    // noinspection JSUnusedGlobalSymbols
    enum Fruits {
      Apple = "apple",
      Banana = "banana",
      Cantaloupe = "cantaloupe",
      A = 5,
    }

    const pickedSchema = z
      .object({
        string: z.string(),
        number: z.number(),
        fixedArrayOfString: z.array(z.string()).nonempty().length(10),
        object: z.object({
          string: z.string(),
        }),
      })
      .partial();

    const circular: z.ZodLazy<z.ZodTypeAny> = z.lazy(() =>
      z.object({
        a: z.string(),
        b: circular,
      })
    );

    const example = z.object({
      string: z.string(),
      number: z.number(),
      arrayOfObjects: z.array(
        z.object({
          string: z.string(),
        })
      ),
      boolean: z.boolean(),
      circular,
      union: z.union([z.object({ number: z.number() }), z.literal("hi")]),
      enum: z.enum(["hi", "bye"]),
      intersectionWithTransform: z
        .number()
        .and(z.bigint())
        .and(z.number().and(z.string()))
        .transform((arg) => console.log(arg)),
      date: z.date(),
      undefined: z.undefined(),
      null: z.null(),
      void: z.void(),
      any: z.any(),
      unknown: z.unknown(),
      never: z.never(),
      optionalString: z.optional(z.string()),
      nullablePartialObject: z.nullable(pickedSchema),
      tuple: z.tuple([
        z.string(),
        z.number(),
        z.object({ string: z.string() }),
      ]),
      record: z.record(
        z.object({
          object: z.object({
            arrayOfUnions: z
              .union([
                z.tuple([z.string(), z.object({ string: z.string() })]),
                z.bigint(),
              ])
              .array(),
          }),
        })
      ),
      map: z.map(z.string(), z.array(z.object({ string: z.string() }))),
      set: z.set(z.string()),
      intersection: z.intersection(z.string(), z.number()).or(z.bigint()),
      promise: z.promise(z.number()),
      function: z
        .function()
        .args(z.string().nullish().default("heo"), z.boolean(), z.boolean())
        .returns(z.string()),
      optDefaultString: z.string().optional().default("hi"),
      refinedStringWithSomeBullshit: z
        .string()
        .refine((val) => val.length > 10)
        .or(z.number())
        .and(z.bigint().nullish().default(1000n)),
      nativeEnum: z.nativeEnum(Fruits),
      lazy: z.lazy(() => z.string()),
      discUnion: z.discriminatedUnion("kind", [
        z.object({ kind: z.literal("circle"), radius: z.number() }),
        z.object({ kind: z.literal("square"), x: z.number() }),
        z.object({ kind: z.literal("triangle"), x: z.number(), y: z.number() }),
      ]),
      branded: z.string().brand("BRAND"),
      catch: z.number().catch(123),
      pipeline: z.string().regex(/\d+/).pipe(z.coerce.number()),
    });

    it("should produce the expected results", () => {
      const node = zodToTs({
        schema: example,
        ...defaultCtx,
      });
      expect(printNode(node)).toMatchSnapshot();
    });
  });

  describe("z.optional()", () => {
    const optionalStringSchema = z.string().optional();
    const objectWithOptionals = z.object({
      optional: optionalStringSchema,
      required: z.string(),
      transform: z
        .number()
        .optional()
        .transform((arg) => arg),
      or: z.number().optional().or(z.string()),
      tuple: z
        .tuple([
          z.string().optional(),
          z.number(),
          z.object({
            optional: z.string().optional(),
            required: z.string(),
          }),
        ])
        .optional(),
    });

    it("outputs correct typescript", () => {
      const node = zodToTs({ schema: optionalStringSchema, ...defaultCtx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    it("should output `?:` and undefined union for optional properties", () => {
      const node = zodToTs({ schema: objectWithOptionals, ...defaultCtx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("z.nullable()", () => {
    const nullableUsernameSchema = z.object({
      username: z.string().nullable(),
    });
    const node = zodToTs({ schema: nullableUsernameSchema, ...defaultCtx });

    it("outputs correct typescript", () => {
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("z.object()", () => {
    it("supports string literal properties", () => {
      const schema = z.object({
        "string-literal": z.string(),
        5: z.number(),
      });
      const node = zodToTs({ schema, ...defaultCtx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    it("does not unnecessary quote identifiers", () => {
      const schema = z.object({
        id: z.string(),
        name: z.string(),
        countryOfOrigin: z.string(),
      });
      const node = zodToTs({ schema, ...defaultCtx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    it("escapes correctly", () => {
      const schema = z.object({
        "\\": z.string(),
        '"': z.string(),
        "'": z.string(),
        "`": z.string(),
        "\n": z.number(),
        $e: z.any(),
        "4t": z.any(),
        _r: z.any(),
        "-r": z.undefined(),
      });
      const node = zodToTs({ schema, ...defaultCtx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    it("supports zod.describe()", () => {
      const schema = z.object({
        name: z.string().describe("The name of the item"),
        price: z.number().describe("The price of the item"),
      });
      const node = zodToTs({ schema, ...defaultCtx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("PrimitiveSchema", () => {
    const primitiveSchema = z.object({
      string: z.string(),
      number: z.number(),
      boolean: z.boolean(),
      date: z.date(),
      undefined: z.undefined(),
      null: z.null(),
      void: z.void(),
      any: z.any(),
      unknown: z.unknown(),
      never: z.never(),
    });
    const node = zodToTs({ schema: primitiveSchema, ...defaultCtx });

    it("outputs correct typescript", () => {
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("z.discriminatedUnion()", () => {
    const shapeSchema = z.discriminatedUnion("kind", [
      z.object({ kind: z.literal("circle"), radius: z.number() }),
      z.object({ kind: z.literal("square"), x: z.number() }),
      z.object({ kind: z.literal("triangle"), x: z.number(), y: z.number() }),
    ]);
    const node = zodToTs({ schema: shapeSchema, ...defaultCtx });

    it("outputs correct typescript", () => {
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("z.literal()", () => {
    test.each([
      z.literal("test"),
      z.literal(true),
      z.literal(false),
      z.literal(123),
    ])("Should produce the correct typescript %#", (schema) => {
      expect(
        printNodeTest(zodToTs({ schema, ...defaultCtx }))
      ).toMatchSnapshot();
    });
  });

  describe("z.effect()", () => {
    describe("transformations", () => {
      test.each([
        { isResponse: false, expected: "intact" },
        { isResponse: true, expected: "transformed" },
      ])("should produce the schema type $expected", ({ isResponse }) => {
        const schema = z.number().transform((num) => `${num}`);
        expect(
          printNodeTest(zodToTs({ schema, ...defaultCtx, isResponse }))
        ).toMatchSnapshot();
      });

      test("should handle unsupported transformation in response", () => {
        const schema = z.number().transform((num) => () => num);
        expect(
          printNodeTest(zodToTs({ schema, ...defaultCtx, isResponse: true }))
        ).toMatchSnapshot();
      });

      test("should handle an error within the transformation", () => {
        const schema = z.number().transform(() => {
          throw new Error("this should be handled");
        });
        expect(
          printNodeTest(zodToTs({ schema, ...defaultCtx, isResponse: true }))
        ).toMatchSnapshot();
      });
    });
  });
});
