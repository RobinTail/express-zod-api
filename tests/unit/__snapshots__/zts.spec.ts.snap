// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zod-to-ts Example should produce the expected results 1`] = `
"{
    string: string;
    number: number;
    arrayOfObjects: {
        string: string;
    }[];
    boolean: boolean;
    circular: any;
    union: {
        number: number;
    } | "hi";
    enum: "hi" | "bye";
    intersectionWithTransform: (number & bigint) & (number & string);
    date: any;
    undefined?: any;
    null: null;
    void?: any;
    any?: any;
    unknown?: any;
    never: any;
    optionalString?: string | undefined;
    nullablePartialObject: {
        string?: string | undefined;
        number?: number | undefined;
        fixedArrayOfString?: string[] | undefined;
        object?: {
            string: string;
        } | undefined;
    } | null;
    tuple: [
        string,
        number,
        {
            string: string;
        }
    ];
    record: Record<string, {
        object: {
            arrayOfUnions: ([
                string,
                {
                    string: string;
                }
            ] | bigint)[];
        };
    }>;
    map: any;
    set: any;
    intersection: (string & number) | bigint;
    promise: any;
    function: any;
    optDefaultString?: string | undefined;
    refinedStringWithSomeBullshit: (string | number) & ((bigint | null) | undefined);
    nativeEnum: "A" | "apple" | "banana" | "cantaloupe" | 5;
    lazy: any;
    discUnion: {
        kind: "circle";
        radius: number;
    } | {
        kind: "square";
        x: number;
    } | {
        kind: "triangle";
        x: number;
        y: number;
    };
    branded: string;
    catch?: number;
    pipeline: string;
}"
`;

exports[`zod-to-ts PrimitiveSchema outputs correct typescript 1`] = `
"{
    string: string;
    number: number;
    boolean: boolean;
    date: any;
    undefined?: any;
    null: null;
    void?: any;
    any?: any;
    unknown?: any;
    never: any;
}"
`;

exports[`zod-to-ts createTypeAlias() optionally takes a comment 1`] = `
"/** A basic user */
type User = {
    username: string;
    age: number;
};"
`;

exports[`zod-to-ts createTypeAlias() outputs correct typescript 1`] = `
"type User = {
    username: string;
    age: number;
};"
`;

exports[`zod-to-ts enums handles numeric literals 1`] = `""Red" | "Green" | "Blue" | 0 | 1 | 2"`;

exports[`zod-to-ts enums handles quoted string literals 1`] = `""Two Words" | "'Quotes\\"" | "\\\\\\"Escaped\\\\\\"" | 0 | 1 | 2"`;

exports[`zod-to-ts enums handles string literals 1`] = `""apple" | "banana" | "cantaloupe""`;

exports[`zod-to-ts z.array() outputs correct typescript 1`] = `
"{
    id: number;
    value: string;
}[]"
`;

exports[`zod-to-ts z.discriminatedUnion() outputs correct typescript 1`] = `
"{
    kind: "circle";
    radius: number;
} | {
    kind: "square";
    x: number;
} | {
    kind: "triangle";
    x: number;
    y: number;
}"
`;

exports[`zod-to-ts z.effect() transformations should handle an error within the transformation 1`] = `"any"`;

exports[`zod-to-ts z.effect() transformations should handle unsupported transformation in response 1`] = `"any"`;

exports[`zod-to-ts z.effect() transformations should produce the schema type intact 1`] = `"number"`;

exports[`zod-to-ts z.effect() transformations should produce the schema type transformed 1`] = `"string"`;

exports[`zod-to-ts z.literal() Should produce the correct typescript 0 1`] = `""test""`;

exports[`zod-to-ts z.literal() Should produce the correct typescript 1 1`] = `"true"`;

exports[`zod-to-ts z.literal() Should produce the correct typescript 2 1`] = `"false"`;

exports[`zod-to-ts z.literal() Should produce the correct typescript 3 1`] = `"123"`;

exports[`zod-to-ts z.nullable() outputs correct typescript 1`] = `
"{
    username: string | null;
}"
`;

exports[`zod-to-ts z.object() does not unnecessary quote identifiers 1`] = `
"{
    id: string;
    name: string;
    countryOfOrigin: string;
}"
`;

exports[`zod-to-ts z.object() escapes correctly 1`] = `
"{
    "\\\\": string;
    "\\"": string;
    "'": string;
    "\`": string;
    "\\n": number;
    $e?: any;
    "4t"?: any;
    _r?: any;
    "-r"?: any;
}"
`;

exports[`zod-to-ts z.object() supports string literal properties 1`] = `
"{
    "5": number;
    "string-literal": string;
}"
`;

exports[`zod-to-ts z.object() supports zod.describe() 1`] = `
"{
    /** The name of the item */
    name: string;
    /** The price of the item */
    price: number;
}"
`;

exports[`zod-to-ts z.optional() outputs correct typescript 1`] = `"string | undefined"`;

exports[`zod-to-ts z.optional() should output \`?:\` and undefined union for optional properties 1`] = `
"{
    optional?: string | undefined;
    required: string;
    transform?: number | undefined;
    or?: (number | undefined) | string;
    tuple?: [
        string | undefined,
        number,
        {
            optional?: string | undefined;
            required: string;
        }
    ] | undefined;
}"
`;
